plugins/inputs/dcos/dcos.go

```diff
@@ -0,0 +1,453 @@
+		if !isItemFiltered(agentContainers, c) {
+			wg.Add(1)
+			go func(cid string) {
+				acc.AddError(m.gatherContainerMetrics(agent.Id, cid, acc))
+				wg.Done()
+			}(c)
+			wg.Add(1)
+			go func(cid string) {
+				acc.AddError(m.gatherContainerAppMetrics(agent.Id, cid, acc))
+				wg.Done()
+			}(c)
+		}
+	}
+	wg.Wait()
+
+	return nil
+}
+
+//getMesosAgentContainers gathers ids of running mesos container
+func (m *Dcos) getMesosAgentContainers(agentId string) ([]string, error) {
+	url := m.ClusterURL + fmt.Sprintf(MesosSlaveContainersUrl, agentId)
+	var agentContainers []mesosAgentContainer
+	err := m.handleJsonRequest(url, &agentContainers)
+	if err != nil {
+		return nil, err
+	}
+	agentContainerIds := make([]string, len(agentContainers))
+	for _, a := range agentContainers {
+		agentContainerIds = append(agentContainerIds, a.Id)
+	}
+	return agentContainerIds, nil
+}
+
+//gatherContainerMetrics collects metric for given container
+func (m *Dcos) gatherContainerMetrics(agentId string, containerId string, acc telegraf.Accumulator) error {
+	url := m.ClusterURL + fmt.Sprintf(AgentContainerMetricsUrl, agentId, containerId)
+	var metric metric
+	err := m.handleJsonRequest(url, &metric)
+	if err != nil {
+		return err
+	}
+	m.processMetric(&metric, acc, Container)
+	return nil
+}
+
+//gatherContainerAppMetrics collects metric for an app in given container, if available
+func (m *Dcos) gatherContainerAppMetrics(agentId string, containerId string, acc telegraf.Accumulator) error {
+	url := m.ClusterURL + fmt.Sprintf(AgentContainerAppMetricsUrl, agentId, containerId)
+	var metric metric
+
+	err := m.handleJsonRequest(url, &metric)
+	if err != nil {
+		return err
+	}
+
+	m.processMetric(&metric, acc, App)
+
+	return nil
+}
+
+//processMetric validated metric and fills accumulator with metric data
+func (m *Dcos) processMetric(metric *metric, acc telegraf.Accumulator, metricType string) {
+
+	measurementData := m.prepareMetric(metric, metricType, acc)
+	//store current timestamp so all measurements for current set of metrics have the same timestamp
+	now := time.Now()
+	for measurementSuffix, tagPoints := range measurementData {
+		for _, points := range tagPoints {
+			tags := make(map[string]string)
+			fields := make(map[string]interface{})
+
+			fillTags(tags, metric.Dimensions)
+
+			tags["scope"] = metricType
+			tags["cluster_url"] = m.ClusterURL
+
+			for _, dp := range points {
+				switch t := dp.Value.(type) {
+				case string: //field values come as real numbers, string is usually "NaN'
+					acc.AddError(fmt.Errorf("Invalid value for field %s: '%s'", dp.Name, t))
+				default:
+					fields[dp.Name] = t
+					fillTags(tags, dp.Tags)
+				}
+			}
+			//fmt.Println(measurementSuffix, fields, tags)
+			acc.AddFields("dcos_"+measurementSuffix, fields, tags, now)
+		}
+	}
+
+}
+
+//fillTags traverses source map and fills tags with non map values
+func fillTags(tags map[string]string, source map[string]interface{}) {
```

Comment:
yes, the main reason is to 'flat' the tags from dimesion, cause sometimes there is the 'labels' sub-tag with other tags. 
I've done it generically, because i don't know exactly what can arrive.

Application metrics tags could be have the same problem, they are not documented